### DISCUSSION TITLE
Support for custom firmware Xiaomi sensors

### DISCUSSION THREAD

**Discussion by mirko3000** [theengs/decoder#489] [New devices/things integration] [answered]
Hello,
just found your amazing app/decoder. Unfortunately it is missing support for the custom firmware for the Xiaomi temperature sensors LYWSD03MMC - see https://github.com/pvvx/ATC_MiThermometer. Would be great if you could add support for one of the custom formats.
Best regards,
Mirko

**Comment by DigiH** [comment]
Hi @mirko3000 

happy to hear that you are enjoying the capabilities of Theengs Decoder. The protocols you request are implemented though, with ATC and PVVX firmwares decoded by the LYWSD03MMC decoder

https://decoder.theengs.io/devices/LYWSD03MMC.html

For decryption of encrypted such protocols there currently only compatible project using Theengs Decoder is Theengs Gateway, the unencrypted versions should be decoded by all other projects though.

May I ask how and with which project you are using Theengs Decoder?

**Comment by mirko3000** [comment]
Hi @DigiH,
great to hear it is supported. I am using the iOS app from the App Store, here it seems that only sensors with the original firmware are being properly detected. For the ones with custom firmware they show up but do not get connected or data gets published.

**Reply by DigiH** [reply]
Hi @mirko3000 

What exactly do you mean by

>  they show up but do not get connected or data gets published.

How do they show up? Looking at the history again of the LYWSD03MMC decoder, I would have assumed that they would already get detected and decoded correctly by the App.

Which of the described firmware do you have installed on your LYWSD03MMCs?
Any special firmware features enabled?

**Reply by mirko3000** [reply]
Hi @DigiH,
See attached screenshot. I have 5 sensors, 3 of them with original firmware, 2 with the custom firmware. I get proper temperature reading from the 3 original, but not from the 2 customs.
Maybe some specific setting needed within the firmware? I tried several advertising formats (including ATC and PVVX), non seem to work.

I have installed the most recent firmware pvvx (https://github.com/pvvx/ATC_MiThermometer), using mostly default values.

![Uploading IMG_C048F5626079-1.jpeg…]()

**Reply by DigiH** [reply]
Hi @mirko3000 

I think something went wrong with the screenshot attachment ;)

A team mate has just tested here with his LYWSD03MMC installing PVVX and it was received correctly in the App, 

I am wondering if the issue stems from you having the same device model with different firmware versions, as the original stock firmware is not being decoded by the Theengs Decoder library, but some legacy decoding within the App, possibly then somehow blocking other same device types to be passed correctly to the Decoder library??

Could you test to see what happens when you take the batteries out of the 3 devices with the original stock firmware and restart the App?

Any possibly try and reattached the screenshot?

**Reply by DigiH** [reply]
Could you also tell us if you had the two LYWSD03MMC, which now have the PVVX firmware, received and entered in the App beforehand when they still had the stock firmware?

**Reply by mirko3000** [reply]
![IMG_C048F5626079-1](https://github.com/theengs/decoder/assets/4167984/6af5de1d-a768-47b5-bbe6-422d9bd42e82)

**Reply by DigiH** [reply, marked as answer]
Thanks @mirko3000 

So it appears that you might have had these two sensors added in the App before when they still had their stock firmware.

Can you try holding your finger on one of the TH Sensor entries, until a green bar appears at the top, then select the trash can icon there to delete this entry. Ideally the device should reappear after a while with it new PVVX firmware being recognised. 

In general the PVVX firmware is preffered to the stock firmware, as the stock firmware's legacy decoding requires a connection to the LYWSD03MMC, which pouts quite a strain on the LYWSD03MMC's battery, causing a greatly shorter battery life. The PVVX firmware decoding however does not affect the battery of the LYWSD03MMC, as it uses passive scanning.

Let's see if you get the correct recognition once you deleted the old entries.

**Reply by mirko3000** [reply]
@DigiH ok that really solved the issue. I removed the devices from the app, once re-appeared data is now showing up correctly!
But I definitely did not have them in the app with stock firmware before, but maybe I modified settings so somehow the got added with non-matching configuration? Anyhow, working now, thanks for your help!

**Reply by DigiH** [reply]
Glad it is working for you now @mirko3000 :) 

I would still suggest to get your other three LYWSD03MMC to the PVVX firmware as well, to have the much better battery life for them, too.

**Reply by mirko3000** [reply]
Will do that, unfortunately the process with latest firmware requires some more effort and does not work OTA anymore :/ But yes, its on my bucket list for the upcoming days!
